std/process.d

```diff
@@ -1050,54 +1050,103 @@ private Pid spawnProcessPosix(scope const(char[])[] args,
-                        foreach (i; 0 .. maxToClose)
+                        DIR* dir = null;
+
+                        // We read from /dev/fd or /proc/self/fd only if the limit is high enough
+                        if (maxDescriptors > 128*1024)
                         {
-                            pfds[i].fd = i + lowfd;
-                            pfds[i].events = 0;
-                            pfds[i].revents = 0;
+                            // Try to open the directory /dev/fd or /proc/self/fd
+                            dir = opendir("/dev/fd");
+                            if (dir is null) dir = opendir("/proc/self/fd");
                         }
-                        if (poll(pfds, maxToClose, 0) >= 0)
+
+                        // If we have a directory, close all file descriptors except stdin, stdout, and stderr
+                        if (dir)
                         {
-                            foreach (i; 0 .. maxToClose)
+                            scope(exit) closedir(dir);
+
+                            int opendirfd = dirfd(dir);
+
+                            // Iterate over all file descriptors
+                            while (true)
                             {
-                                // POLLNVAL will be set if the file descriptor is invalid.
-                                if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                                dirent* entry = readdir(dir);
+
+                                if (entry is null) break;
+                                if (entry.d_name[0] == '.') continue;
+
+                                int fd = atoi(cast(char*) entry.d_name);
+
+                                // Don't close stdin, stdout, stderr, or the directory file descriptor
+                                if (fd < lowfd || fd == opendirfd) continue;
+
+                                close(fd);
                             }
                         }
                         else
                         {
-                            // Fall back to closing everything.
-                            foreach (i; lowfd .. maxDescriptors)
+                            // This is going to allocate 8 bytes for each possible file descriptor from lowfd to r.rlim_cur
+                            if (maxDescriptors <= 128*1024)
                             {
-                                close(i);
+                                // NOTE: malloc() and getrlimit() are not on the POSIX async
+                                // signal safe functions list, but practically this should
+                                // not be a problem. Java VM and CPython also use malloc()
+                                // in its own implementation via opendir().
+                                import core.stdc.stdlib : malloc;
+                                import core.sys.posix.poll : pollfd, poll, POLLNVAL;
+
+                                immutable maxToClose = maxDescriptors - lowfd;
```

Comment:
This could (only theoretically) be a problem because if maxDescriptors were equal to 4, the result would be 0. But I doubt there are cases where the ulimit is set to 4. In that case probably `pds` on next lines will trigger an InternalError. 

I think we can safely ignore this case, just a note. If not we can add a if (maxToClose > 0) for the next block.

Reply:
I think it works, because the point is to keep the first 3 handles open. lowfd is an int, maxDescriptors is an int, so the result will not wrap.

I believe foreach on an inverted range will just not execute, so we are fine here.

Reply:
Not the foreach, but the malloc. But again, who cares. If they have just 4 as max fd probably they have some other bigger problems :)